pkg/mountexecutor/mountexecutor.go

```diff
@@ -0,0 +1,126 @@
+	Mounter    *backend.SnapshotMounter
+}
+
+// MountOptions are options for a single mount request
+type MountOptions struct {
+	// Context here is only valid for synchronous mounts
+	Context          context.Context
+	NamedRef         docker.Named
+	VolumeId         string
+	TargetPath       string
+	VolumeCapability *csi.VolumeCapability
+	ReadOnly         bool
+}
+
+// MountExecutor executes mount
+type MountExecutor struct {
+	asyncMount bool
+	mutex      *sync.Mutex
+	mounter    *backend.SnapshotMounter
+	asyncErrs  map[docker.Named]error
+}
+
+// NewMountExecutor initializes a new mount executor
+func NewMountExecutor(o *MountExecutorOptions) *MountExecutor {
+	return &MountExecutor{
+		asyncMount: o.AsyncMount,
+		mutex:      &sync.Mutex{},
+		mounter:    o.Mounter,
+	}
+}
+
+// StartMounting starts the mounting
+func (m *MountExecutor) StartMounting(o *MountOptions) error {
+
+	if pullstatus.Get(o.NamedRef) != pullstatus.Pulled || mountstatus.Get(o.NamedRef) == mountstatus.StillMounting {
+		return nil
+	}
+
+	ro := o.ReadOnly ||
+		o.VolumeCapability.AccessMode.Mode == csi.VolumeCapability_AccessMode_SINGLE_NODE_READER_ONLY ||
+		o.VolumeCapability.AccessMode.Mode == csi.VolumeCapability_AccessMode_MULTI_NODE_READER_ONLY
+
+	if !m.asyncMount {
+		mountstatus.Update(o.NamedRef, mountstatus.StillMounting)
+		if err := m.mounter.Mount(o.Context, o.VolumeId, backend.MountTarget(o.TargetPath), o.NamedRef, ro); err != nil {
+			mountstatus.Update(o.NamedRef, mountstatus.Errored)
+			return err
+		}
+		mountstatus.Update(o.NamedRef, mountstatus.Mounted)
+		return nil
+	}
+
+	go func() {
+		m.mutex.Lock()
+		ctx, cancel := context.WithTimeout(context.Background(), time.Minute*10)
+		defer cancel()
+
+		if mountstatus.Get(o.NamedRef) == mountstatus.StillMounting {
+			m.mutex.Unlock()
+			return
+		}
+
+		mountstatus.Update(o.NamedRef, mountstatus.StillMounting)
+		if err := m.mounter.Mount(ctx, o.VolumeId, backend.MountTarget(o.TargetPath), o.NamedRef, ro); err != nil {
+			klog.Errorf("mount err: %v", err.Error())
+			mountstatus.Update(o.NamedRef, mountstatus.Errored)
+			m.asyncErrs[o.NamedRef] = fmt.Errorf("err: %v: %v", err, m.asyncErrs[o.NamedRef])
+			m.mutex.Unlock()
+			return
+		}
+		mountstatus.Update(o.NamedRef, mountstatus.Mounted)
+		m.mutex.Unlock()
+	}()
+
+	return nil
+}
+
+// WaitForMount waits for the volume to get mounted
+func (m *MountExecutor) WaitForMount(o *MountOptions) error {
+	if pullstatus.Get(o.NamedRef) != pullstatus.Pulled {
+		return nil
+	}
+
+	if !m.asyncMount {
+		return nil
+	}
+
+	mountCondFn := func() (done bool, err error) {
+		if mountstatus.Get(o.NamedRef) == mountstatus.Mounted {
+			return true, nil
+		}
+		if m.asyncErrs[o.NamedRef] != nil {
+			return false, m.asyncErrs[o.NamedRef]
+		}
+		return false, nil
+	}
+
+	if err := wait.PollImmediate(
+		100*time.Millisecond,
+		2*time.Minute,
+		mountCondFn); err != nil {
+		return fmt.Errorf("waited too long to mount the image: %v", err)
```

Comment:
We already do it [here](https://github.com/warm-metal/csi-driver-image/pull/71/files/a6130917d17896cfdcea675ad3e990b0a6744d86#diff-acd03f9f72e4eaf35379ca3f76525f4964b3709422f8059fa6dbf16206b6f141R86)

Reply:
Yes, but it seems like it's in a different function `StartMounting()`. Do we also want it in `WaitforMount()` function as it's called separately?

Reply:
I don't think we need it in `WaitForMount` because are just watching for errors which occur in `StartMounting`. Updating `mountstatus` in `WaitForMount` on error would mean updating `mountstatus` with error twice (once in `StartMounting` and once in `WaitForMount`)